### ISSUE TITLE
[no-unnecessary-type-assertion] does not flag redundant 'as const' on literals in   contextually-typed positions

### ISSUE THREAD

**Issue by alexkahndev** [typescript-eslint/typescript-eslint#12308]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-type-assertion/

### Description

Opening this per the suggestion on #12297.

I ran into a redundant as const in a real codebase that the rule didn't catch, and when I went to fix it I noticed there used to be a literal // TODO - add contextually unnecessary check for this comment in this rule covering the case. It was removed in #11789 without being implemented, so I wanted to flag it before assuming it was intentional.

The case is as const on a primitive literal in a position where TypeScript's contextual type is already a literal (or a union of literals) that accepts the value. For example:

In both cases the as const is pure noise: removing it leaves the inferred type identical because contextual typing already preserves the literal. The rule today only flags as const in declarations like const x = 'a' as const (via the checkLiteralConstAssertions option from #10979), so these contextual cases slip through.

### Fail

```TypeScript
type Recipient = { kind: 'a' | 'b' };
const r: Recipient = { kind: 'a' as const }; // 'as const' does nothing here

declare function f(x: 'a' | 'b'): void;
f('a' as const); // same — contextual type already narrows it
```

### Pass

```TypeScript
type Recipient = { kind: 'a' | 'b' };
const r: Recipient = { kind: 'a' };

declare function f(x: 'a' | 'b'): void;
f('a' );
```

**Comment by bradzacher** [comment]
I don't know if this is something that we can do. There are so many cases where TS can change the type of the underlying literal itself in response to the `as const`. Similarly there are cases of generics that behave in similar ways. Which can mean that when inspecting the types it can look like a useless assertion but it's actually required.

I'm happy if someone wants to research and build something out but we'll need some extensive tests to ensure we won't regress anything.

**Comment by alexkahndev** [comment]
I put up #12317. Your generic case turned out to be a real false positive in my first cut: `getContextualType` returns the post-inference type, so `id<T>('a' as const)` looks like it has a literal contextual type precisely because of the assertion the rule would ask you to remove. The PR takes a conservative path and bails on any argument to a generic call (any callee whose signatures declare type parameters). Tests cover the generic patterns explicitly: `id<T>(x: T), arrayOf<T>(...items: T[])`, and `box<T>(v: T)` without target annotations as valid, plus `arr.push` on `('a' | 'b')[]` as invalid to prove the rule still fires when T is pinned by the receiver.
